### PR TITLE
Cursor aware Component Browser

### LINKED ISSUE #5220: Finish Component Browser suggestions based on the cursor position
This task is automatically imported from the old Task Issue Board and it was originally created by Adam Obuchowicz.
Original issue is [here](https://www.pivotaltracker.com/story/show/184181435).
<hr />

Finish the task #183521918. The branch should be rebased, cleaned up and ready for being taken over on 10 Jan.

**QA:** @farmaazon

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

Closes #5220

This PR implements the cursor-aware behavior of the CB, as described in [pivotal issue](https://www.pivotaltracker.com/n/projects/2539304/stories/183521918)

https://user-images.githubusercontent.com/6566674/221807206-39f93cb4-8253-421d-a33a-33ac0aa56e54.mp4

https://user-images.githubusercontent.com/6566674/223124947-259153ca-e656-4349-87b5-47c06fd21af2.mp4

### Important Notes

- The `intended_method` of the node's metadata is marked as deprecated, and all usages are removed.
- It seems *all usages of Scala parser are removed from IDE*. We no longer use it to parse documentation for snippets.

This is how the snippets docs look now:

Screenshot 2023-02-28 at 13 18 11

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: app/gui/src/controller/searcher.rs [hunk 1]
```diff
@@ -628,7 +439,8 @@ impl Searcher {
         let def_id = graph.graph().id;
         let def_span = double_representation::module::definition_span(&module_ast, &def_id)?;
         let module_repr: Rope = module_ast.repr().into();
-        let position = module_repr.offset_to_location_snapped(def_span.end);
+        // TODO: is it correct?
+        let position = module_repr.offset_to_location_snapped(def_span.end - text::ByteDiff(1));
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher.rs]: Subtracting one byte from position helps to find local variables declared inside `main` function. Apparently the engine thinks that locals are out of scope otherwise and doesn't send them in `completion` response. This issue should be properly investigated separately.

File: app/gui/src/controller/searcher.rs [hunk 2]
```diff
@@ -725,24 +537,24 @@ impl Searcher {
...
-        if expression_changed {
+    pub fn set_input(&self, new_input: String, cursor_position: Byte) -> FallibleResult {
+        debug!("Manually setting input to {new_input} with cursor position {cursor_position}");
+        let parsed_input = input::Input::parse(self.ide.parser(), new_input, cursor_position);
+        let new_context = parsed_input.context().map(|ctx| ctx.into_ast().repr());
+        let old_input = mem::replace(&mut self.data.borrow_mut().input, parsed_input);
+        let old_context = old_input.context().map(|ctx| ctx.into_ast().repr());
+
+        self.invalidate_picked_suggestions();
+        let context_changed = old_context != new_context;
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher.rs:563]: I'm not sure if we're doing the correct thing here. Previous code compared expressions, now we compare access chains, and they usually `None`. So we are reloading the list far less frequently now.

File: app/gui/src/controller/searcher.rs [hunk 3]
```diff
@@ -2209,64 +1711,57 @@ pub mod test {
...
+        let new_input = searcher.use_suggestion(suggestion1.clone()).unwrap();
+        assert_eq!(new_input.text, "test_var_1 ");
+        let new_input = searcher.use_suggestion(suggestion1).unwrap();
+        assert_eq!(new_input.text, "test_var_1 ");
+        let (function, arg1, arg2) = picked_suggestions().iter().cloned().expect_tuple();
+        assert!(are_same(&function.entry, &test_function_1));
+        assert!(are_same(&arg1.entry, &test_var_1));
+        assert!(are_same(&arg2.entry, &test_var_1));
+
+        // TODO
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher.rs]: I don't understand what's going on in this part of the test, I will fix it or remove it.

File: app/gui/src/controller/searcher/component/hardcoded.rs [hunk 4]
```diff
@@ -106,10 +91,8 @@ impl Snippet {
     /// documentation parser cannot be created or the argument fails to parse as valid
     /// documentation.
     fn with_documentation(mut self, documentation: &str) -> Self {
-        let doc_parser = parser_scala::DocParser::new().unwrap();
-        let doc_string = documentation.to_string();
-        let documentation_html = doc_parser.generate_html_doc_pure(doc_string);
-        self.documentation_html = Some(documentation_html.unwrap());
+        // TODO: we need the Scala parser in searcher tests only for HTML documentation.
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher/component/hardcoded.rs]: This doesn't look like the correct change, but removing the doc parser usage allows to turn all searcher tests into native ones. This is a huge advantage while developing, though I probably should revert this afterwards and make searcher tests `wasm_test` again.

File: app/gui/src/controller/searcher/input.rs [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+        let code_to_insert = suggestion.code_to_insert(generate_this);
+        debug!("Code to insert: \"{code_to_insert}\"");
+        let end_of_inserted_code = replaced.start + text::Bytes(code_to_insert.len());
+        let end_of_inserted_text = end_of_inserted_code + text::Bytes(1);
+        let mut new_input = self.ast.to_string();
+        let raw_range = replaced.start.value..replaced.end.value;
+        let range_start = raw_range.start;
+        let range_end = raw_range.end;
+        Self::ensure_on_char_boundary(&new_input, range_start)?;
+        Self::ensure_on_char_boundary(&new_input, range_end)?;
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher/input.rs]: I saw rare panics when one of the indices was not on char boundary inside the input, which was weird. This code helps to mitigate the issue if it ever happens again.

File: app/gui/src/model/module.rs [hunk 6]
```diff
@@ -411,6 +411,7 @@ pub struct NodeMetadata {
     ///
     /// The methods may be defined for different types, so the name alone don't specify them.
     #[serde(default, deserialize_with = "enso_prelude::deserialize_or_default")]
+    // TODO: It is not used anywhere. Deprecate?
     pub intended_method: Option<MethodId>,
```
Comment (vitvakatu) [inline app/gui/src/model/module.rs:415]: Now all usages of the `intended_method` are removed. We are no longer using it to communicate with the engine, so this field should be probably removed or marked as deprecated.

File: app/gui/src/controller/searcher/component/hardcoded.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/src/controller/searcher/component/hardcoded.rs]: At the same time, the documentation for  snippets is extremely primitive and doc parser doesn't really bring much.

File: app/gui/src/controller/searcher.rs (hunk 1 above)
Comment (farmaazon) [reply app/gui/src/controller/searcher.rs]: Let's make a GH issue for that and put its number here.

File: app/gui/src/controller/searcher/component/hardcoded.rs (hunk 4 above)
Comment (farmaazon) [reply app/gui/src/controller/searcher/component/hardcoded.rs]: We can stay with this change. Before the searcher, tests were usually wasm_bindgen anyway, because we used the Scala parser. But I guess it's no longer the case?

Comment (vitvakatu) [reply app/gui/src/controller/searcher/component/hardcoded.rs]: Yep, now the tests use Rust parser in general.

File: app/gui/src/controller/searcher.rs (hunk 1 above)
Comment (vitvakatu) [reply app/gui/src/controller/searcher.rs]: I fixed the issue completely by searching position before creating a new node.

File: app/gui/src/controller/searcher.rs (hunk 3 above)
Comment (vitvakatu) [reply app/gui/src/controller/searcher.rs]: Seems it is not compatible with the new behavior, removed this piece

File: app/gui/src/controller/searcher/component/hardcoded.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/src/controller/searcher/component/hardcoded.rs]: I decided to remove the scala parser from here, it does not bring us much.

File: app/gui/src/model/module.rs (hunk 6 above)
Comment (vitvakatu) [reply app/gui/src/model/module.rs:415]: Marked it as deprecated and removed all usages from the code.

File: app/gui/src/controller/searcher/input.rs [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+            if edited_ast_on_right.edited_name.is_some() {
+                edited_ast_on_right
+            } else {
+                edited_ast_on_left
+            }
+        } else {
+            edited_ast_on_left
+        };
+        let ast = InputAst::Line(ast);
+        Self { ast, cursor_position, edited_ast }
```
Comment (farmaazon) [inline app/gui/src/controller/searcher/input.rs:290]: This code deserves a bit of explanation in the comment.

File: app/gui/src/controller/searcher/input.rs (hunk 5 above)
Comment (farmaazon) [reply app/gui/src/controller/searcher/input.rs]: I'd "inline" the `range_start` and `range_end` variables. They won't make the lines too long.

File: app/gui/src/controller/searcher/input.rs [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+    }
+}
+
+#[derive(Clone, Debug, Fail)]
+#[allow(missing_docs)]
+#[fail(display = "Not a char boundary: index {} at '{}'.", index, string)]
+pub struct NotACharBoundary {
+    index:  usize,
+    string: String,
+}
```
Comment (farmaazon) [inline app/gui/src/controller/searcher/input.rs]: Let's put it in `Errors` section somewhere at the top.

File: app/gui/src/controller/graph.rs [hunk 9]
```diff
@@ -999,14 +1013,8 @@ impl Handle {
...
-        let metadata = self.module.node_metadata(id).ok()?;
-        let db_entry = db.lookup_method(metadata.intended_method?)?;
-        // If the name is different than intended method than apparently it is not intended anymore
-        // and should be ignored.
-        let matching = if let Some(name) = name { name == db_entry.name } else { true };
-        matching.then(|| db_entry.invocation_info())
+    fn call_info(&self, _id: node::Id, _name: Option<&str>) -> Option<CalledMethodInfo> {
+        None
     }
 }
```
Comment (farmaazon) [inline app/gui/src/controller/graph.rs]: We already have `Empty` context for such cases: https://github.com/enso-org/enso/blob/d8f6f3fd52d87f63d8b067b609162423ed23ebcc/app/gui/language/span-tree/src/generate/context.rs

So, if we are not in some weird generics' machinery, the controller should not implement Context at all, and use `Empty` instead if itself where needed.

File: app/gui/src/controller/searcher.rs (hunk 2 above)
Comment (farmaazon) [reply app/gui/src/controller/searcher.rs:563]: This is correct. Before we were filtering the components by expected argument type (and it depended on the left side of the expression), but now the filtering depends only on accessor chain (i.e. the context).

File: app/gui/src/test.rs [hunk 10]
```diff
@@ -264,7 +264,7 @@ pub mod mock {
             let mut json_client = language_server::MockClient::default();
             let mut binary_client = binary::MockClient::new();
             // Creating a searcher controller always triggers a query for completion.
-            controller::searcher::test::expect_completion(&mut json_client, &[]);
+            // controller::searcher::test::expect_completion(&mut json_client, &[]);
```
Comment (farmaazon) [inline app/gui/src/test.rs]: Commented code. Remove if unneeded.

File: app/gui/view/graph-editor/src/component/node/input/area.rs [hunk 11]
```diff
@@ -850,6 +852,8 @@ ensogl::define_endpoints! {
         /// Set the node expression.
         set_expression (node::Expression),
 
+        edit_expression (text::Range<Byte>, ImString),
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/component/node/input/area.rs:861]: Missing docs.

**Review by farmaazon — APPROVED** [review]

File: app/gui/src/controller/searcher/input.rs (hunk 8 above)
Comment (vitvakatu) [reply app/gui/src/controller/searcher/input.rs]: Done

File: app/gui/src/controller/graph.rs (hunk 9 above)
Comment (vitvakatu) [reply app/gui/src/controller/graph.rs]: Done, also removed a couple of tests that relied on  arguments information from node metadata.

File: app/gui/src/test.rs (hunk 10 above)
Comment (vitvakatu) [reply app/gui/src/test.rs]: Fixed it.

File: app/gui/src/controller/searcher.rs [hunk 12]
```diff
@@ -971,25 +714,22 @@ impl Searcher {
...
+        }
 
         let node_id = self.mode.node_id();
-        let input_chain = self.data.borrow().input.as_prefix_chain(self.ide.parser());
-        let expression = self.get_expression(input_chain);
-        let intended_method = self.intended_method();
+        let expression = self.get_expression(self.data.borrow().input.ast());
         self.graph.graph().set_expression(node_id, expression)?;
-        if let Mode::NewNode { .. } = self.mode.as_ref() {
+        let graph = self.graph.graph();
```
Comment (kazcw) [inline app/gui/src/controller/searcher.rs:763]: Is this bound for convenience, or to capture the value returned by `graph()` before a subsequent operation changes it? If it's for convenience it should be used consistently so it doesn't look like something tricky is happening; if it is intentionally capturing an older value, a comment would be helpful.

File: app/gui/src/controller/searcher/input.rs [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+            new_input,
+            replaced,
+            inserted_code: (replaced.start..end_of_inserted_code).into(),
+            inserted_text: (replaced.start..end_of_inserted_text).into(),
+        })
+    }
+
+    /// Return `true` if the `index` is on a char boundary inside `string` and can be used as a
+    /// range boundary. Otherwise, return `false` and report an error.
+    fn ensure_on_char_boundary(string: &str, index: usize) -> FallibleResult<()> {
```
Comment (kazcw) [inline app/gui/src/controller/searcher/input.rs]: Docs don't match return type.

File: app/gui/src/controller/searcher/input.rs [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+        }
+    }
+
+    /// Return the part of the input used as filtering pattern.
+    pub fn pattern(&self) -> &str {
+        if let Some(edited) = &self.edited_ast.edited_name {
+            let cursor_position_in_name = self.cursor_position - edited.range.start;
+            let name = ast::identifier::name(&edited.ast);
+            name.map_or("", |n| {
+                &n[text::Range::new(text::Byte(0), text::Byte(0) + cursor_position_in_name)]
```
Comment (kazcw) [inline app/gui/src/controller/searcher/input.rs]: This slice can panic if `edited_ast` and `cursor_position` are inconsistent. That *shouldn't* happen, but the invariant relied on is not trivial--slicing with `get` here would reduce the consequences of potential future bugs.

**Review by kazcw — APPROVED** [review]
Looks good. @farmaazon I think this answers my questions about how the frontend uses the Scala doc parser!

**Comment by farmaazon** [comment]
# QA report

## Issues

When starting editing node in the middle of identifier, I see CB with all entries first, and the filtering appears only after a moment.

https://user-images.githubusercontent.com/3919101/222146581-a07528ec-7efd-4f0e-8baa-eed9711ff277.mp4

## Bad UX

Besides the above issue, the PR meets the acceptance criteria. Still, when testing, I found additional things worsening my UX:

* When writing literals (text, number), the results are no longer filtered, so we consistently try to add the first selected entry (usually Column.from_vector) to the input. Before, because literals were part of pattern (and usually no entry matched the literal) the component browser was empty. The solution would be to not display suggestions when the cursor is in or next to a literal, or, even better, we could display a single entry, for example `"foo bar"` if user puts `"foo bar`, or `43` if user types `43` - but this is worth a separate task.
* When editing existing static method invocation, accepting the same method adds the type name again, resulting in erroneous code. It may also be seen as error in visualization preview:

Screenshot from 2023-03-01 13-50-13

Solution for it is implementing #5356 we may consider keeping this branch unmerged until we implement and merge that task to it.

**Comment by vitvakatu** [comment]
In commit https://github.com/enso-org/enso/pull/5770/commits/e4b606e394e61f854d5c58334aeacd25f412404c I implemented a requested feature: when writing literals we display a single entry, for example "foo bar" if user puts `"foo bar`, or `43` if user types `43`.


https://user-images.githubusercontent.com/6566674/223124947-259153ca-e656-4349-87b5-47c06fd21af2.mp4

**Review by farmaazon — COMMENTED** [review]
The code changes for the new "Literal" entry looks good.

**Review by wdanilo — APPROVED** [review]
I don't have other code-related comments than Adam. However, I've got a few questions - each starts with "do we support it / when we will support it":
- not putting space after replacing a variable. E.g. we have `foo| + bar`, where `|` is mouse cursor. After replacing `foo` with `foo2` we want to see `foo2 |+ bar`.
- Supporting string literals with multiple starting and ending quotes.
- Proper support for negative number literals.

**Comment by vitvakatu** [comment]
> not putting space after replacing a variable. E.g. we have foo| + bar, where | is mouse cursor. After replacing foo with foo2 we want to see foo2 |+ bar.

Tracked in https://github.com/enso-org/enso/issues/5848

>    Supporting string literals with multiple starting and ending quotes.

Already implemented. `'''`, `"""`, `'`, `"` are supported as delimiters. Showing documentation for `Text` and `Number` types is also implemented.
    
> Proper support for negative number literals.

Tracked in #5826

**Comment by wdanilo** [comment]
@vitvakatu add `good first issue` label for issues that you think are simple to be implemented and someone from the community could try implementing. The linked issues look like those to me, but I might be wrong.

**Review by JaroslavTulach — APPROVED** [review]

**Comment by wdanilo** [comment]
@vitvakatu what is the status of the PR? It was accepted 2 weeks ago (!), and the QA was done then. Why is it not merged yet?

**Comment by vitvakatu** [comment]
@wdanilo as we agreed with @farmaazon, we will merge #5356 into this PR, and then merge the whole thing. The QA done 2 weeks ago discovered a few issues, and they were addressed during that two weeks.

**Comment by wdanilo** [comment]
@vitvakatu gotcha! Sorry, I forgot about it. Please make short notes about agreements in the future so we will not be missing them. Thanks again! :)

**Comment by wdanilo** [comment]
@vitvakatu as we are making changes here, could you please move it to draft stage and request review again after the code is final, please?

**Comment by vitvakatu** [comment]
QA was done in #5960 on an identical state of the code.
